### ISSUE TITLE
 521.wrf_r segfaults when built with -flto=full -O2

### ISSUE THREAD

**Issue by bryanpkc** [flang-compiler/flang#1429]
After upgrading to LLVM 17, the 521.wrf_r benchmark in SPEC CPU2017 is miscompiled at O3 with LTO enabled. Examination of the crash site shows that a large amount of code that used to be inlined into `wrf_init` (via `alloc_and_configure_domain`) has been deleted, leaving the `head_grid` pointer uninitialized.

The root cause of this problem is that upstream LLVM had strengthened the pruning of unreachable code to handle `undef` and `poison`. Specifically, these three patches are found to be relevant:

- https://github.com/llvm/llvm-project/commit/1fc425380e
- https://github.com/llvm/llvm-project/commit/70aca7b122
- https://github.com/llvm/llvm-project/commit/41895843b5

Classic Flang leaves variables uninitialized by default. The corresponding IR values are therefore considered to contain `undef`. In our downstream compiler, we have tested with initializing all variables to zeroes, and that successfully avoids the problem.

628.pop2_s is also affected by this problem.

**Comment by bryanpkc** [comment]
As discussed at the call, I will follow up in this issue with information about which variables are left uninitialized in the source, and whether the problem occurs at `-O2`/`-O3`.

**Comment by bryanpkc** [comment]
The problem does occur at `-O2` and `-O3`.

**Comment by bryanpkc** [comment]
The problematic `undef` values that caused some basic blocks to be pruned correspond to the dummy arguments `ipex` and `jpex` in the function `compute_memory_dims_rsl_lite` (`module_dm.F90`, line 1088). This function is called by `patch_domain_rsl_lite`, which passes a whole bunch of actual arguments, none of which is initialized before the call.

**Comment by bryanpkc** [comment]
@pawosm-arm @shivaramaarao Could you check things on your side to see if you can reproduce the problem, or determine why you don't see the problem?

**Comment by pawosm-arm** [comment]
Hi @bryanpkc assuming the problem manifests itself at runtime, I need to ask, what input data set are you using? (test, ref, or something else).

**Comment by bryanpkc** [comment]
@pawosm-arm I have confirmed that the problem can be reproduced without specifying any `-mcpu=` option. The run-time segmentation fault happens during benchmark start-up; I am using the `test` data set.

**Comment by pawosm-arm** [comment]
Been testing it all night in our CI, with your flags, with O3 and even with Ofast, and it didn't fail. It's time to do it manually, but bear in mind, building spec's wrf with LTO take ages, and I want to try two different builds of the compiler, so it may take days before I'd confirm anything.

**Comment by pawosm-arm** [comment]
Just noticed, our CI always adds `-funroll-loops` to the explicitly specified set of flags, but I doubt it matters

**Comment by pawosm-arm** [comment]
...also we don't seem to use (or even build) lld, but it also should make no difference on what LTO does when optimizing the code

**Comment by pawosm-arm** [comment]
Should `628.pop2_s` also fail during benchmark start-up? It takes forever to execute it, so I guess it's not affected in my builds...

**Comment by bryanpkc** [comment]
@pawosm-arm Can you confirm in the `521.wrf_r` source files that the actual arguments I mentioned are not initialized? I am using an older version of the benchmark (v1.0.5) and perhaps SPEC has changed the source since then?

**Comment by shivaramaarao** [comment]
[AMD Official Use Only - General]

I could reproduce the issue with -O2 and -flto. I need to still check why we don’t see the issue in our downstream compiler. Will update when I have additional information.

Regards,
Shivaram

**Comment by pawosm-arm** [comment]
> Since I can't reproduce the problem with the most recent release of our compiler, I'll try to build spec2k17's wrf with vanilla classic.

Indeed, wrf fails when built like that (spec2k17 ver. 1.1.5), but pop2 doesn't.

**Comment by bryanpkc** [comment]
As I mentioned above, our downstream compiler has an option to zero-initialize variables automatically, so we have a workaround, which we may be able to upstream. gfortran has [a similar set of options](https://gcc.gnu.org/onlinedocs/gfortran/Code-Gen-Options.html#index-finit-local-zero). Oracle's Fortran compiler has an `-xcheck=init_local` option that initializes variables to trapping values, [for debugging purpose](https://docs.oracle.com/cd/E19059-01/stud.10/819-0492/3_options.html#pgfId-270661). ifort has a similar [`-check uninit` option](https://www.intel.com/content/www/us/en/docs/fortran-compiler/developer-guide-reference/2023-2/check.html).

I had a quick look at the language standard. While it discusses definition and undefinition of variables, it does not specify any default initialization behaviour, and uses of uninitialized variables are undefined behaviour. From what I can see in online discussions, programmers are encouraged to define variables explicitly.

What should be our next step? Any opinion? @pawosm-arm @shivaramaarao

**Comment by pawosm-arm** [comment]
I'm slightly confused. Doesn’t classic-flang already zero-initialize variables (at least globals/module variables)?

**Comment by bryanpkc** [comment]
I think the global/module variables are allocated in the BSS section and zero-initialized only as a side effect. Local variables that are not explicitly initialized are left undefined.
